File: 프로그래머스/Lv1/실패율.js
// 내 풀이
function solution(N, stages) {
    let answer = [];
    let level = {};
    let level_ratio = {};
    let total = stages.length;
    
    for(let i=1; i<N+1; i++){
        let cnt = stages.filter((x) => i === x);
        level[String(i)] = cnt.length;
    }
    
    for(let i in level){
        level_ratio[String(i)] = level[i] / total;
        total -= level[i];
    }
    
    level_ratio = Object.entries(level_ratio).sort((a, b) => b[1] - a[1]);
    
    for(let i=0; i<level_ratio.length; i++){
        answer.push(parseInt(level_ratio[i][0]));
    }
    
    return answer;
}

// 다른 사람 풀이
function solution2(N, stages){
    let answer = [];

    for(let i=1; i<=N; i++){
        let cnt = stages.filter((x) => x === i).length;
        let ratio = stages.filter((x) => x >= i).length;

        answer.push([i, cnt/ratio]);
    }

    answer.sort((a, b) => b[1] - a[1]);

    return answer.map((i) => i[0]);
}

console.log(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]));
console.log(solution2(4, [4,4,4,4,4]));

File: 프로그래머스/Lv1/실패율.py
# 내 풀이
def solution(N, stages):
    answer = []
    
    for i in range(1, N+1):
        cnt = stages.count(i)
        ratio = len([x for x in stages if x >= i])
        
        if cnt == 0:
            answer.append([i, 0])
        else:
            answer.append([i, cnt/ratio])
        
    return [i[0] for i in sorted(answer, key=lambda x:x[1], reverse=True)]

# 다른 사람 풀이
def solution2(N, stages):
    answer = []
    fail = []
    info = [0] * (N + 2)
    print(info)
    for stage in stages:
        info[stage] += 1
    for i in range(N):
        be = sum(info[(i + 1):]) # 레벨에 도달했고 다 푼 사람
        yet = info[i + 1] # 레벨에 도달했지만 못 푼 사람
        if be == 0:
            fail.append((str(i + 1), 0))
        else:
            fail.append((str(i + 1), yet / be))
    for item in sorted(fail, key=lambda x: x[1], reverse=True):
        answer.append(int(item[0]))
    return answer

print(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]))
print(solution2(4, [4,4,4,4,4]))